integration-tests/tests/common/mod.rs

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! This crate contains the common code for integration tests
+
+#![deny(clippy::all)]
+#![deny(missing_docs)]
+
+use std::path::PathBuf;
+use std::process::Command;
+
+pub struct Setup {
+    pub wksp_root: PathBuf,
+    pub keep: PathBuf,
+    pub payload_dir: PathBuf,
+    pub shim: PathBuf,
+}
+
+impl Setup {
+    pub fn new() -> Self {
+        Self {
+            wksp_root: PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+                .parent()
+                .unwrap()
+                .to_path_buf(),
+            keep: find_filepath(
+                (PathBuf::from(env!("CARGO_MANIFEST_DIR")).parent().unwrap())
+                    .join("enarx-keep-sgx/target/")
+                    .to_path_buf(),
+                "enarx-keep-sgx",
+            )
+            .unwrap(),
+            payload_dir: (PathBuf::from(env!("CARGO_MANIFEST_DIR")).parent().unwrap())
+                .join("integration-tests/target/x86_64-unknown-linux-musl/")
+                .to_path_buf(),
+            shim: find_filepath(
+                (PathBuf::from(env!("CARGO_MANIFEST_DIR")).parent().unwrap())
+                    .join("enarx-keep-sgx-shim/target/x86_64-unknown-linux-musl/")
+                    .to_path_buf(),
+                "enarx-keep-sgx-shim",
+            )
+            .unwrap(),
+        }
+    }
+}
```

Comment:
This solution does improve code reuse, but we can take it a couple of steps further.

1. We can more seamlessly generate code that works for all of our keeps (not just sgx) using conditional compilation. I'm thinking of a free-standing function like this:

```rust
#[cfg_attr(not(any(has_sgx, has_sev)), ignore)]
pub const fn keep_under_test() -> &'static str {
    #[cfg(feature = "sgx")]
    "enarx-keep-sgx"
    
    #[cfg(feature = "sev")]
    "enarx-keep-sev"
}
```

can help us fill in some of the setup code in a more flexible way without having to change any of our code:

```rust
            keep: find_filepath(
                (PathBuf::from(env!("CARGO_MANIFEST_DIR")).parent().unwrap())
                    .join(format!("{}/target/", keep_under_test())
                    .to_path_buf(),
                keep_under_test(),
            )
```

furthermore, since we've been consistently naming our crates `enarx-keep-blah` and `enarx-keep-blah-shim` this would let us do:

```rust
            shim: find_filepath(
                (PathBuf::from(env!("CARGO_MANIFEST_DIR")).parent().unwrap())
                    .join(format!("{}-shim/target/x86_64-unknown-linux-musl/", keep_under_test()))
                    .to_path_buf(),
                format!("{}-shim", keep_under_test()),
            )
```

There are lots of calls to `PathBuf::from(env!("CARGO_MANIFEST_DIR")).parent().unwrap()`, I'd suggest factoring them all out and storing this as a variable at the beginning of the function.

As far as selecting the payload directory:

```rust             
payload_dir: (PathBuf::from(env!("CARGO_MANIFEST_DIR")).parent().unwrap())
                .join("integration-tests/target/x86_64-unknown-linux-musl/")
                .to_path_buf(),
```

I think this is an artifact from these binaries living in the `payload` crate in a previous revision. This could probably be closer to

```rust
payload_dir: PathBuf::from(env!("CARGO_MANIFEST_DIR"))
                .join("target/x86_64-unknown-linux-musl/")
                .to_path_buf(),
```

After all these changes, the `Setup` struct is beginning to look like a constructor function for a `Command`, which we can use to avoid having to instantiate a `Command` struct by hand in each test.

I think `struct Setup` should be changed to `struct IntegrationTest(Command)`.

Then its constructor function could end with something like:

```rust
pub fn new(bin: &str) -> Self {
    // snipped setup
    // "bin" is the name of the payload we find during the setup stage,
    // ultimately its full filepath should wind up in the `code` variable referred to below

    let cmd = Command::new(keep)
        .current_dir(wksp_root)
        .arg("--shim")
        .arg(shim)
        .arg("--code")
        .arg(code)

    Self(cmd)
}
```

Then it can offer a method called something like:

```rust
pub fn exec(&self, timeout: Option<Duration>) -> ExitStatus {
    let timeout = time.unwrap_or(Duration::from_secs(60));
    let status = self.0.spawn().unwrap();
    match status.wait_timeout(timeout).unwrap() {
        None => {
            self.0.kill.unwrap();
            panic!("killed by watchdog");
        },
        Some(exit_status) => exit_status
       }
    }
}
```

So after all that, ideally running an integration test should be as easy as:

```rust
#[test]
fn test_sys_exit() {
    let test = IntegrationTest::new("test_sys_exit");
    assert!(test.exec(None).success());
}
```

Regardless of what keep we are trying to test. I think the above suggestions should accommodate most of our integration testing needs for now.

Also just a disclaimer that I haven't tried compiling _any_ of my code snippets above; I'm just thinking out loud. So if you decide to incorporate some of the ideas above, some elbow grease might be needed to keep them to compile :slightly_smiling_face: